File: python/diamond/diamond.py
"""Generate a Diamond pattern."""

import string


def rows(letter: str) -> list[str]:
    """Return a diamond pattern for a letter."""
    count = string.ascii_uppercase.index(letter)
    line_width = count * 2 + 1
    lines = []
    for offset in range(0, count + 1):
        char = string.ascii_uppercase[offset]
        center = " " * (offset * 2 - 1)
        line = center.join(char * (2 if offset else 1))
        lines.append(f"{line:^{line_width}}")

    # Mirror the top half for the bottom.
    return lines + lines[-2::-1]

File: go/diamond/diamond.go
package diamond

import (
	"bytes"
	"fmt"
)

// A is the value of 'A'.
const A = int('A')

// Gen returns a diamond pattern string.
func Gen(char byte) (string, error) {
	if char < 'A' || char > 'Z' {
		return "", fmt.Errorf("invalid char %b not in range [A, Z]", char)
	}

	length := int(char) - A
	rowLength := (length * 2) + 1
	blankRow := bytes.Repeat([]byte{' '}, rowLength)

	// Precompute all the rows, since we need them twice.
	rows := make([][]byte, rowLength)
	for i := 0; i <= length; i++ {
		// Fill the row with spaces.
		rows[i] = make([]byte, rowLength)
		copy(rows[i], blankRow)
		// Insert the char on either side of center.
		rows[i][length-i] = byte(A + i)
		rows[i][length+i] = byte(A + i)
	}

	var buf bytes.Buffer

	// Write the top half.
	for i := 0; i < length; i++ {
		buf.Write(rows[i])
		buf.WriteByte('\n')
	}
	// Write the bottom half.
	for i := length; i > 0; i-- {
		buf.Write(rows[i])
		buf.WriteByte('\n')
	}
	// Final row, no newline.
	buf.Write(rows[0])

	return buf.String(), nil
}
